File: packages/autoTest/src/types/JSBCommand.ts
import { BackgroundMaterial, CornerRadius, Vec3, WindowStyle } from './types'

// Base command interface, similar to Swift's CommandDataProtocol
interface CommandDataProtocol {
  commandType: string
}

// Spatial object command interface
interface SpatialObjectCommand extends CommandDataProtocol {
  id: string
}

// Spatialized element properties interface
interface SpatializedElementProperties extends SpatialObjectCommand {
  name?: string
  clientX?: number
  clientY?: number
  width?: number
  height?: number
  depth?: number
  backOffset?: number
  rotationAnchor?: Vec3
  opacity?: number
  visible?: boolean
  scrollWithParent?: boolean
  zIndex?: number
  enableDragStart?: boolean
  enableDragGesture?: boolean
  enableDragEndGesture?: boolean
  enableRotateStartGesture?: boolean
  enableRotateGesture?: boolean
  enableRotateEndGesture?: boolean
  enableMagnifyStartGesture?: boolean
  enableMagnifyGesture?: boolean
  enableMagnifyEndGesture?: boolean
  enableTapGesture?: boolean
}

// Update spatial scene properties command
export class UpdateSpatialSceneProperties implements CommandDataProtocol {
  commandType = 'UpdateSpatialSceneProperties'
  id?: string
  material?: BackgroundMaterial
  cornerRadius?: CornerRadius
  opacity?: number
}

// Create spatial scene command
export class CreateSpatialScene implements CommandDataProtocol {
  commandType = 'CreateSpatialScene'
  id?: string
  name?: string
  version?: string
  properties?: Record<string, any>
}

// Add spatialized element to spatial scene command
export class AddSpatializedElementToSpatialScene
  implements CommandDataProtocol
{
  commandType = 'AddSpatializedElementToSpatialScene'
  spatializedElementId: string = ''
}

// Inspect spatial scene command
export class InspectSpatialScene implements CommandDataProtocol {
  commandType = 'InspectSpatialScene'
  id?: string
}

// Create spatialized 2D element command
export class CreateSpatialized2DElement implements CommandDataProtocol {
  commandType = 'CreateSpatialized2DElement'
  id: string = ''
  url: string = ''
}

// Update spatialized 2D element properties command
export class UpdateSpatialized2DElementProperties
  implements SpatializedElementProperties
{
  commandType = 'UpdateSpatialized2DElementProperties'
  id: string = ''
  name?: string
  clientX?: number
  clientY?: number
  width?: number
  height?: number
  depth?: number
  backOffset?: number
  rotationAnchor?: Vec3
  opacity?: number
  visible?: boolean
  scrollWithParent?: boolean
  zIndex?: number
  enableDragStart?: boolean
  enableDragGesture?: boolean
  enableDragEndGesture?: boolean
  enableRotateStartGesture?: boolean
  enableRotateGesture?: boolean
  enableRotateEndGesture?: boolean
  enableMagnifyStartGesture?: boolean
  enableMagnifyGesture?: boolean
  enableMagnifyEndGesture?: boolean
  enableTapGesture?: boolean
  scrollPageEnabled?: boolean
  material?: BackgroundMaterial
  cornerRadius?: CornerRadius
}

// Update spatialized element transform command
export class UpdateSpatializedElementTransform implements SpatialObjectCommand {
  commandType = 'UpdateSpatializedElementTransform'
  id: string = ''
  matrix?: number[] // add matrix property to match Swift definition
}

// Create spatialized static 3D element command
export class CreateSpatializedStatic3DElement implements CommandDataProtocol {
  commandType = 'CreateSpatializedStatic3DElement'
  modelURL: string = ''
}

// Create spatialized dynamic 3D element command
export class CreateSpatializedDynamic3DElement implements CommandDataProtocol {
  commandType = 'CreateSpatializedDynamic3DElement'
  test: boolean = false
}

// Create spatial entity command
export class CreateSpatialEntity implements CommandDataProtocol {
  commandType = 'CreateSpatialEntity'
  name?: string
}

// Create geometry properties command
export class CreateGeometryProperties implements CommandDataProtocol {
  commandType = 'CreateGeometry'
  type: string = ''
  width?: number
  height?: number
  depth?: number
  cornerRadius?: number
  splitFaces?: boolean
  radius?: number
}

// Create unlit material command
export class CreateUnlitMaterial implements CommandDataProtocol {
  commandType = 'CreateUnlitMaterial'
  color?: string
  textureId?: string
  transparent?: boolean
  opacity?: number
}

// Create texture command
export class CreateTexture implements CommandDataProtocol {
  commandType = 'CreateTexture'
  url: string = ''
}

// Create model asset command
export class CreateModelAsset implements CommandDataProtocol {
  commandType = 'CreateModelAsset'
  url: string = ''
}

// Create spatial model entity command
export class CreateSpatialModelEntity implements CommandDataProtocol {
  commandType = 'CreateSpatialModelEntity'
  modelAssetId: string = ''
  name?: string
}

// Create model component command
export class CreateModelComponent implements CommandDataProtocol {
  commandType = 'CreateModelComponent'
  geometryId: string = ''
  materialIds: string[] = []
}

// Add component to entity command
export class AddComponentToEntity implements CommandDataProtocol {
  commandType = 'AddComponentToEntity'
  entityId: string = ''
  componentId: string = ''
}

// Add entity to dynamic 3D command
export class AddEntityToDynamic3D implements CommandDataProtocol {
  commandType = 'AddEntityToDynamic3D'
  dynamic3dId: string = ''
  entityId: string = ''
}

// Add entity to entity command
export class AddEntityToEntity implements CommandDataProtocol {
  commandType = 'AddEntityToEntity'
  childId: string = ''
  parentId: string = ''
}

// Set parent for entity command
export class SetParentForEntity implements CommandDataProtocol {
  commandType = 'SetParentToEntity'
  childId: string = ''
  parentId?: string
}

// Remove entity from parent command
export class RemoveEntityFromParent implements CommandDataProtocol {
  commandType = 'RemoveEntityFromParent'
  entityId: string = ''
}

// Update entity properties command
export class UpdateEntityProperties implements CommandDataProtocol {
  commandType = 'UpdateEntityProperties'
  entityId: string = ''
  transform: Record<string, number> = {}
}

// Update entity event command
export class UpdateEntityEvent implements CommandDataProtocol {
  commandType = 'UpdateEntityEvent'
  type: string = ''
  entityId: string = ''
  isEnable: boolean = false
}

// Convert from entity to entity command
export class ConvertFromEntityToEntity implements CommandDataProtocol {
  commandType = 'ConvertFromEntityToEntity'
  fromEntityId: string = ''
  toEntityId: string = ''
  position: Vec3
  constructor() {
    this.position = { x: 0, y: 0, z: 0 }
  }
}

// Convert from entity to scene command
export class ConvertFromEntityToScene implements CommandDataProtocol {
  commandType = 'ConvertFromEntityToScene'
  fromEntityId: string = ''
  position: Vec3
  constructor() {
    this.position = { x: 0, y: 0, z: 0 }
  }
}

// Convert from scene to entity command
export class ConvertFromSceneToEntity implements CommandDataProtocol {
  commandType = 'ConvertFromSceneToEntity'
  entityId: string = ''
  position: Vec3
  constructor() {
    this.position = { x: 0, y: 0, z: 0 }
  }
}

// Keep original Inspect class for compatibility
export class Inspect implements CommandDataProtocol {
  commandType = 'Inspect'
  id: string = ''
}

// Destroy command
export class DestroyCommand implements CommandDataProtocol {
  commandType = 'Destroy'
  id: string = ''
}

// Update spatialized static 3D element properties command
export class UpdateSpatializedStatic3DElementProperties
  implements SpatializedElementProperties
{
  commandType = 'UpdateSpatializedStatic3DElementProperties'
  id: string = ''
  name?: string
  clientX?: number
  clientY?: number
  width?: number
  height?: number
  depth?: number
  backOffset?: number
  rotationAnchor?: Vec3
  opacity?: number
  visible?: boolean
  scrollWithParent?: boolean
  zIndex?: number
  enableDragStart?: boolean
  enableDragGesture?: boolean
  enableDragEndGesture?: boolean
  enableRotateStartGesture?: boolean
  enableRotateGesture?: boolean
  enableRotateEndGesture?: boolean
  enableMagnifyStartGesture?: boolean
  enableMagnifyGesture?: boolean
  enableMagnifyEndGesture?: boolean
  enableTapGesture?: boolean
  modelURL?: string
  modelTransform?: number[]
}

// Update spatialized dynamic 3D element properties command
export class UpdateSpatializedDynamic3DElementProperties
  implements SpatializedElementProperties
{
  commandType = 'UpdateSpatializedDynamic3DElementProperties'
  id: string = ''
  name?: string
  clientX?: number
  clientY?: number
  width?: number
  height?: number
  depth?: number
  backOffset?: number
  rotationAnchor?: Vec3
  opacity?: number
  visible?: boolean
  scrollWithParent?: boolean
  zIndex?: number
  enableDragStart?: boolean
  enableDragGesture?: boolean
  enableDragEndGesture?: boolean
  enableRotateStartGesture?: boolean
  enableRotateGesture?: boolean
  enableRotateEndGesture?: boolean
  enableMagnifyStartGesture?: boolean
  enableMagnifyGesture?: boolean
  enableMagnifyEndGesture?: boolean
  enableTapGesture?: boolean
}

// Add spatialized element to spatialized 2D element command
export class AddSpatializedElementToSpatialized2DElement
  implements SpatialObjectCommand
{
  commandType = 'AddSpatializedElementToSpatialized2DElement'
  id: string = ''
  spatializedElementId: string = ''
}

// Baseplate visibility enum
export enum BaseplateVisibilityJSB {
  automatic = 'automatic',
  visible = 'visible',
  hidden = 'hidden',
}

// World scaling enum
export enum WorldScalingJSB {
  automatic = 'automatic',
  dynamic = 'dynamic',
}

// World alignment enum
export enum WorldAlignmentJSB {
  adaptive = 'adaptive',
  automatic = 'automatic',
  gravityAligned = 'gravityAligned',
}

// Size interface
interface Size {
  width: number
  height: number
}

// Resize range interface
interface ResizeRange {
  min: Size
  max: Size
}

// Scene options JSB interface
export class XSceneOptionsJSB {
  defaultSize?: Size
  type?: WindowStyle
  resizability?: ResizeRange
  worldScaling?: WorldScalingJSB
  worldAlignment?: WorldAlignmentJSB
  baseplateVisibility?: BaseplateVisibilityJSB
}

// Update scene config command
export class UpdateSceneConfigCommand implements CommandDataProtocol {
  commandType = 'UpdateSceneConfig'
  config: XSceneOptionsJSB
  constructor(data: XSceneOptionsJSB) {
    this.config = data
  }
}

// Focus scene command
export class FocusSceneCommand implements CommandDataProtocol {
  commandType = 'FocusScene'
  id: string
  constructor(id: string) {
    this.id = id
  }
}

// Get spatial scene state command
export class GetSpatialSceneStateCommand implements CommandDataProtocol {
  commandType = 'GetSpatialSceneState'
}

// Export base interfaces for other files to use
export {
  CommandDataProtocol,
  SpatialObjectCommand,
  SpatializedElementProperties,
}

File: packages/visionOS/web-spatial/JSBCommand.swift
import SwiftUI

struct UpdateSpatialSceneProperties: CommandDataProtocol {
    static let commandType: String = "UpdateSpatialSceneProperties"
    let cornerRadius: CornerRadius?
    let material: BackgroundMaterial?
    let opacity: Double?
}

struct AddSpatializedElementToSpatialScene: CommandDataProtocol {
    static let commandType: String = "AddSpatializedElementToSpatialScene"
    let spatializedElementId: String
}

struct CreateSpatializedStatic3DElement: CommandDataProtocol {
    static let commandType: String = "CreateSpatializedStatic3DElement"
    let modelURL: String
}

struct CreateSpatializedDynamic3DElement: CommandDataProtocol {
    static let commandType: String = "CreateSpatializedDynamic3DElement"
    let test: Bool
}

struct CreateSpatialEntity: CommandDataProtocol {
    static let commandType: String = "CreateSpatialEntity"
    let name: String?
}

struct CreateGeometryProperties: CommandDataProtocol {
    static let commandType: String = "CreateGeometry"
    let type: String
    let width: Float?
    let height: Float?
    let depth: Float?
    let cornerRadius: Float?
    let splitFaces: Bool?
    let radius: Float?
}

struct CreateUnlitMaterial: CommandDataProtocol {
    static let commandType: String = "CreateUnlitMaterial"
    let color: String?
    let textureId: String?
    let transparent: Bool?
    let opacity: Float?
}

struct CreateTexture: CommandDataProtocol {
    static let commandType: String = "CreateTexture"
    let url: String
}

struct CreateModelAsset: CommandDataProtocol {
    static let commandType: String = "CreateModelAsset"
    let url: String
}

struct CreateSpatialModelEntity: CommandDataProtocol {
    static let commandType: String = "CreateSpatialModelEntity"
    let modelAssetId: String
    let name: String?
}

struct CreateModelComponent: CommandDataProtocol {
    static let commandType: String = "CreateModelComponent"
    let geometryId: String
    let materialIds: [String]
}

struct AddComponentToEntity: CommandDataProtocol {
    static let commandType: String = "AddComponentToEntity"
    let entityId: String
    let componentId: String
}

struct AddEntityToDynamic3D: CommandDataProtocol {
    static let commandType: String = "AddEntityToDynamic3D"
    let dynamic3dId: String
    let entityId: String
}

struct AddEntityToEntity: CommandDataProtocol {
    static let commandType: String = "AddEntityToEntity"
    let childId: String
    let parentId: String
}

struct SetParentForEntity: CommandDataProtocol {
    static let commandType: String = "SetParentToEntity"
    let childId: String
    let parentId: String?
}

struct RemoveEntityFromParent: CommandDataProtocol {
    static let commandType: String = "RemoveEntityFromParent"
    let entityId: String
}

struct UpdateEntityProperties: CommandDataProtocol {
    static let commandType: String = "UpdateEntityProperties"
    let entityId: String
    let transform: [String: Float]
}

struct UpdateEntityEvent: CommandDataProtocol {
    static let commandType: String = "UpdateEntityEvent"
    let type: String
    let entityId: String
    let isEnable: Bool
}

struct ConvertFromEntityToEntity: CommandDataProtocol {
    static let commandType: String = "ConvertFromEntityToEntity"
    let fromEntityId: String
    let toEntityId: String
    let position: Vec3
}

struct ConvertFromEntityToScene: CommandDataProtocol {
    static let commandType: String = "ConvertFromEntityToScene"
    let fromEntityId: String
    let position: Vec3
}

struct ConvertFromSceneToEntity: CommandDataProtocol {
    static let commandType: String = "ConvertFromSceneToEntity"
    let entityId: String
    let position: Vec3
}

struct InspectCommand: CommandDataProtocol {
    static let commandType: String = "Inspect"
    var id: String?
}

protocol SpatialObjectCommand: CommandDataProtocol {
    var id: String { get }
}

struct DestroyCommand: CommandDataProtocol {
    static let commandType: String = "Destroy"
    var id: String
}

protocol SpatializedElementProperties: SpatialObjectCommand {
    var name: String? { get }
    var clientX: Double? { get }
    var clientY: Double? { get }
    var width: Double? { get }
    var height: Double? { get }
    var depth: Double? { get }
    var backOffset: Double? { get }
    var rotationAnchor: Vec3? { get }
    var opacity: Double? { get }
    var visible: Bool? { get }
    var scrollWithParent: Bool? { get }
    var zIndex: Double? { get }

    var enableDragStartGesture: Bool? { get }
    var enableDragGesture: Bool? { get }
    var enableDragEndGesture: Bool? { get }

    var enableRotateGesture: Bool? { get }
    var enableRotateEndGesture: Bool? { get }
    var enableMagnifyGesture: Bool? { get }
    var enableMagnifyEndGesture: Bool? { get }
    var enableTapGesture: Bool? { get }
}

struct UpdateSpatialized2DElementProperties: SpatializedElementProperties {
    static let commandType: String = "UpdateSpatialized2DElementProperties"
    let id: String
    let name: String?
    var clientX: Double?
    var clientY: Double?
    let width: Double?
    let height: Double?
    let depth: Double?
    let backOffset: Double?
    let rotationAnchor: Vec3?
    let opacity: Double?
    let visible: Bool?
    let scrollWithParent: Bool?
    let zIndex: Double?

    var enableDragStartGesture: Bool?
    var enableDragGesture: Bool?
    var enableDragEndGesture: Bool?
    var enableRotateGesture: Bool?
    var enableRotateEndGesture: Bool?
    var enableMagnifyGesture: Bool?
    var enableMagnifyEndGesture: Bool?
    var enableTapGesture: Bool?

    let scrollPageEnabled: Bool?
    let material: BackgroundMaterial?
    let cornerRadius: CornerRadius?

    /// this value is used by previous WebSpatial code, keep it here only for Compatibility consideration
    /// may delete it when we think it's not needed
    let scrollEdgeInsetsMarginRight: Double?
}

struct UpdateSpatializedStatic3DElementProperties: SpatializedElementProperties {
    static let commandType: String = "UpdateSpatializedStatic3DElementProperties"
    let id: String
    let name: String?
    var clientX: Double?
    var clientY: Double?
    let width: Double?
    let height: Double?
    let depth: Double?
    let backOffset: Double?
    let rotationAnchor: Vec3?
    let opacity: Double?
    let visible: Bool?
    let scrollWithParent: Bool?
    let zIndex: Double?

    var enableDragStartGesture: Bool?
    let enableDragGesture: Bool?
    let enableDragEndGesture: Bool?
    let enableRotateGesture: Bool?
    let enableRotateEndGesture: Bool?
    let enableMagnifyGesture: Bool?
    let enableMagnifyEndGesture: Bool?
    let enableTapGesture: Bool?

    let modelURL: String?
    let modelTransform: [Double]?
}

struct UpdateSpatializedDynamic3DElementProperties: SpatializedElementProperties {
    static let commandType: String = "UpdateSpatializedDynamic3DElementProperties"
    let id: String
    let name: String?
    var clientX: Double?
    var clientY: Double?
    let width: Double?
    let height: Double?
    let depth: Double?
    let backOffset: Double?
    let rotationAnchor: Vec3?
    let opacity: Double?
    let visible: Bool?
    let scrollWithParent: Bool?
    let zIndex: Double?

    var enableDragStartGesture: Bool?
    let enableDragGesture: Bool?
    let enableDragEndGesture: Bool?
    let enableRotateGesture: Bool?
    let enableRotateEndGesture: Bool?
    let enableMagnifyGesture: Bool?
    let enableMagnifyEndGesture: Bool?
    let enableTapGesture: Bool?
}

struct UpdateSpatializedElementTransform: SpatialObjectCommand {
    static let commandType: String = "UpdateSpatializedElementTransform"
    let id: String
    let matrix: [Double]
}

struct AddSpatializedElementToSpatialized2DElement: SpatialObjectCommand {
    static let commandType: String = "AddSpatializedElementToSpatialized2DElement"
    let id: String
    let spatializedElementId: String
}

/// incomming JSB data
struct XSceneOptionsJSB: Codable {
    let defaultSize: Size?
    let type: SpatialScene.WindowStyle?
    let resizability: ResizeRange?
    let worldScaling: WorldScalingJSB?
    let worldAlignment: WorldAlignmentJSB?
    let baseplateVisibility: BaseplateVisibilityJSB?
}

enum BaseplateVisibilityJSB: String, Codable {
    case automatic
    case visible
    case hidden

    var toSDK: Visibility {
        switch self {
        case .automatic: return .automatic
        case .visible: return .visible
        case .hidden: return .hidden
        }
    }
}

enum WorldScalingJSB: String, Codable {
    case automatic
    case dynamic

    var toSDK: WorldScalingBehavior {
        switch self {
        case .automatic: return .automatic
        case .dynamic: return .dynamic
        }
    }
}

enum WorldAlignmentJSB: String, Codable {
    case adaptive
    case automatic
    case gravityAligned

    var toSDK: WorldAlignmentBehavior {
        switch self {
        case .adaptive: return .adaptive
        case .automatic: return .automatic
        case .gravityAligned: return .gravityAligned
        }
    }
}

struct UpdateSceneConfigCommand: CommandDataProtocol {
    static let commandType = "UpdateSceneConfig"
    let config: XSceneOptionsJSB
    init(_ data: XSceneOptionsJSB) {
        config = data
    }
}

struct FocusSceneCommand: CommandDataProtocol {
    static let commandType = "FocusScene"
    let id: String
    init(_ id: String) {
        self.id = id
    }
}

struct GetSpatialSceneStateCommand: CommandDataProtocol {
    static let commandType = "GetSpatialSceneState"
}

struct UpdateAttachmentEntityCommand: CommandDataProtocol {
    static let commandType = "UpdateAttachmentEntity"
    let id: String
    let position: [Float]?
    let size: AttachmentSize?
}

struct AttachmentSize: Codable {
    let width: Double
    let height: Double
}
